peps/pep-0740.rst

```diff
@@ -302,63 +322,36 @@ object) by selecting a new version number.
 
 .. _payload-and-signature-generation:
 
-Attestation payload and signature generation
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-
-The *attestation payload* is the actual claim that is cryptographically signed
-over within the attestation object (as the ``message_signature``).
-
-The attestation payload is encoded as an :rfc:`8785` canonicalized JSON object,
-with the following pseudocode layout:
-
-.. code-block:: python
-
-  @dataclass
-  class AttestationPayload:
-      distribution: str
-      """
-      The file name of the Python package distribution.
-      """
-
-      digest: str
-      """
-      The SHA-256 digest of the distribution's contents, as a hexadecimal string.
-      """
-
-The value of ``distribution`` is the same distribution filename that appears
-in the :pep:`503` and :pep:`691` APIs. For example, ``distribution`` would be
-``sampleproject-1.2.0-py2.py3-none-any.whl`` for the following simple index
-entry:
-
-.. code-block:: html
-
-  <a href="https://example.com/...">sampleproject-1.2.0-py2.py3-none-any.whl</a><br/>
-
-In practice, this means that ``distribution`` is defined by the PyPA's
-living specifications for
-:ref:`binary distributions <packaging:binary-distribution-format>` and
-:ref:`source distributions <packaging:source-distribution-format>`, although
-non-conforming distributions may be hosted by the index.
+Attestation statement and signature generation
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-The following pseudocode demonstrates the construction of an attestation
-payload and its signature:
+The *attestation statement* is the actual claim that is cryptographically signed
+over within the attestation object (i.e., the ``envelope.statement``).
 
-.. code-block:: python
+The attestation statement is encoded as a
+`v1 in-toto Statement object <https://github.com/in-toto/attestation/blob/v1.0/spec/v1.0/statement.md>`__,
+in JSON form. When serialized the statement is treated as an opaque binary blob,
+avoiding the need for canonicalization. An example JSON-encoded statement is
+provided in :ref:`appendix-4`.
 
-  def build_payload(dist: Path) -> AttestationPayload:
-      return AttestationPayload(
-        distribution=dist.name,
-        digest=sha256(dist.read_bytes()).hexdigest,
-      )
+In addition to being a v1 in-toto Statement, the attestation statement is constrained
+in the following ways:
 
-  attestation_payload = build_payload("sampleproject-1.2.0-py2.py3-none-any.whl")
+* The in-toto ``subject`` **MUST** contain only a single subject.
+* ``subject[0].name`` is the distribution's filename, which **MUST** be
+  a valid :ref:`source distribution <packaging:source-distribution-format>` or
+  :ref:`wheel distribution <packaging:binary-distribution-format>` filename.
```

Comment:
I think this is the best we can do in the general case: ideally we'd fully "ultranormalize" each filename here by e.g. sorting compressed tag sets, but in practice there's too much backend diversity for this. We could mandate it as part of adopting these features, but that too will cause too much heartache IMO (since the point of changing this PEP to allow in-toto attestations is to make life easier for people who want to use GitHub's attestation feature with a minimum number of changes).